recipes/LJSpeech/quantization/train.py

```diff
@@ -1,142 +1,267 @@
+
+        # Extract features
+        with torch.no_grad():
+            self.modules.ssl_model.eval()
+            feats = self.modules.ssl_model(sig, lens)  # [K, B, N, H]
+            feats = feats[self.hparams.layer_id]  # [B, N, H]
+
+        return feats
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the objectives."""
+        feats = predictions  # [B, N, H]
+
+        if stage != sb.Stage.TRAIN:
+            # For K-means the validation/test loss is the inertia
+            # The lower the inertia, the better should be the clustering
+            # It is useful to monitor progress across epochs
+            # However, when saving checkpoints we always keep the last one (i.e. max_keys=["epoch"])
+            # to keep backward compatibility
+            loss = self.hparams.quantizer.inertia(feats)
+            return loss
+
+        # If training, accumulate features (batch size used for K-means training
+        # should be much larger than batch size used for feature extraction)
+        feats = feats.flatten(end_dim=-2)  # [BN, H]
+        self.curr_feats.append(feats)
+        self.curr_batch_size += len(feats)
+        if self.curr_batch_size < self.hparams.kmeans_batch_size:
+            # If not enough features, leave average loss unchanged and go to next batch
+            # avg_loss is computed as: (avg_loss - avg_loss / self.step) + float(loss) / self.step
+            # If we set loss = avg_loss, avg_loss stays unchanged
+            loss = torch.tensor(self.avg_train_loss)
+            # Keep compatibility with standard supervised training
+            # (SpeechBrain expects a tensor with gradient)
+            loss.requires_grad_()
+            return loss
+        self.curr_feats = torch.cat(self.curr_feats)
+        feats = self.curr_feats[: self.hparams.kmeans_batch_size]
+
+        # Keep remaining features for next iteration
+        self.curr_feats = [self.curr_feats[self.hparams.kmeans_batch_size :]]
+        self.curr_batch_size = len(self.curr_feats[0])
 
-logger = get_logger(__name__)
+        # Retrieve current centroids
+        old_cluster_centers = self.hparams.quantizer.cluster_centers
+
+        # Partial fit on current batch
+        self.hparams.quantizer.partial_fit(feats)
+
+        # For K-means the training loss is the drift between current centroids and old centroids
+        # If close to 0, it means that the training has converged
+        curr_cluster_centers = self.hparams.quantizer.cluster_centers
+        loss = (curr_cluster_centers - old_cluster_centers).norm()
+
+        # Keep compatibility with standard supervised training
+        # (SpeechBrain expects a tensor with gradient)
+        loss.requires_grad_()
+        self.optimizer_step += 1
+        assert self.optimizer_step == self.modules.quantizer.n_steps, (
+            f"optimizer_step: {self.optimizer_step}",
+            f"quantizer.n_steps: {self.modules.quantizer.n_steps}",
+        )
```

Comment:
remark: I actually do think that SpeechBrain Brain class will need at some point a small refactor to allow the possibility to return a Loss object that is not from PyTorch and where you won't do BP on it.